### PR TITLE
Implement dash pattern scaling in pdf

### LINKED ISSUE #6588: Different line styles between PNG and PDF exports.
### Is the line style supposed to depend on the export backend?
_MPL 1.5.1 is from Anaconda and MPL 2.x is from Github, both on Linux (CentOS 7)_

With MPL 1.5.1, I get the same line styles between both PDF and PNG exports:
- PNG @ 100 DPI: 

mpl_ver_1 5 1_100dpi
- PDF @ None DPI: [MPL_ver_1.5.1_None.pdf](https://github.com/matplotlib/matplotlib/files/316429/MPL_ver_1.5.1_None.pdf)

while with master version, they are different between PDF and PNG exports:
- PNG @ 100 DPI: 

mpl_ver_2 0 0b1 post1628 g1aba7c3_100dpi
- PDF @ None DPI: [MPL_ver_2.0.0b1.post1628+g1aba7c3_None.pdf](https://github.com/matplotlib/matplotlib/files/316430/MPL_ver_2.0.0b1.post1628.g1aba7c3_None.pdf)

NB: the problem seems to be independent from the DPI value: I get similar line styles between both PNG exports with 100 DPI and 300 DPI.
- PNG @ 300 DPI: 

mpl_ver_2 0 0b1 post1628 g1aba7c3_300

The example snippet is
```python
import matplotlib
import matplotlib.pyplot as plt
import numpy as np

test_case = matplotlib.__version__

fig, ax = plt.subplots(num=test_case, figsize=(4.8, 3.2))
ax.set_title("MPL version: {0:s}".format(test_case))

xx = np.array([0.0, 1.0])

ax.plot(xx, 1 * xx, ls='dashed', label='default lw')
ax.plot(xx, 2 * xx, ls='dashed', lw=2, label='lw=2')
ax.plot(xx, 3 * xx, ls='dashed', lw=3, label='lw=3')

ax.legend(loc='upper left')

plt.show()

for ext, dpi in (('pdf', None), ('png', 100), ('png', 300)):
    fig.savefig("MPL_ver_{v:s}_{d}.{e:s}".format(v=test_case, d=dpi, e=ext),
                dpi=dpi)
```

My guess would that I may be somehow related to the new behavior introduced in #5926, _but I didn't have time to have a look yet_.

### PR DESCRIPTION
**Pull request by jkseppan**
Factor the dash pattern scaling into a separate function so `GraphicsContextPdf.delta` can call it. Combine the linewidth and dash comparisons because now linewidth affects dashing.

Fixes #6588 

I didn't see if this breaks anything else, it's likely to change the test results for all tests with dashed lines.

### CODE REVIEW

**Comment by tacaswell** [comment]
This might also address the pdf related issue brought up in #5430

**Comment by afvincent** [comment]
@tacaswell About your comment on #5430 , are you sure there is an issue with the PDF backend and the offset negative values? Because the produced PDF file looks fine to me, with and without jkseppan's fix (I rapidly copy-pasted his fix and re-tested it on #5430 MWE).

**Comment by tacaswell** [comment]
@afvincent Yes, sorry 2nd time being wrong about that :disappointed:

**Comment by jkseppan** [comment]
I don't think this patch does anything for #5430. It would be a small change, though.

**Comment by afvincent** [comment]
Wouldn't it be worth including @jkseppan's handling of negative dash offsets into `backend_bases.py` (for example in the method `scale_dashes`) instead of just in `backend_pdf.py`? It might solve the remaining issue in #5430 with the PNG export.

**Comment by mdboom** [comment]
@jkseppan: This looks good and I'd like to get this in for the next release candidate.  Would you mind updating the test images so everything passes?

**Comment by tacaswell** [comment]
The lw related computation probably needs to be pulled all they way up into `Line2D` to fix https://github.com/matplotlib/matplotlib/issues/6592

**Comment by jkseppan** [comment]
@mdboom: I added new baseline images, but now I'm not so sure this fixes the difference between  backends. If you compare the files `lib/matplotlib/tests/baseline_images/test_axes/dash_offset.png` and `lib/matplotlib/tests/baseline_images/test_axes/dash_offset.pdf` the pdf file has much longer dashes.

File: lib/matplotlib/backends/backend_pdf.py [hunk 1]
```diff
@@ -2206,15 +2206,19 @@ def capstyle_cmd(self, style):
...
+    def linewidth_dash_cmd(self, width, dashes):
+        result = []
+        if self.get_linewidth() != width:
+            result += [width, Op.setlinewidth]
+        if self._dashes != dashes:
+            offset, dash = self.scale_dashes(width, dashes)
+            if offset is None or dash is None:
+                dash = []
+                offset = 0
+            if sum(dash) > 0:
```
Comment (tacaswell) [inline lib/matplotlib/backends/backend_pdf.py]: This check should probably be moved up into the base class?

**Comment by tacaswell** [comment]
Something is not right that this is leaking into the tests under the classic style.

I have an idea of how to fix this and #6592, but am not completely sure it will work.

**Pull request closed without merging**